src/Featurizers/StringFeaturizer.h

```diff
@@ -88,9 +137,85 @@ class StringEstimator : public Components::InferenceOnlyEstimatorImpl<StringTran
 // |
 // ----------------------------------------------------------------------
+template <typename T>
+StringTransformer<T>::StringTransformer(bool useEmptyStringsForNullValues /*=false*/) :
+    _useEmptyStringsForNullValues(
+        std::move(
+            [&useEmptyStringsForNullValues](void) -> bool & {
+                if(useEmptyStringsForNullValues && Traits<T>::IsNullableType == false)
+                    throw std::invalid_argument("empty strings cannot be used with types that are not nullable");
+
+                return useEmptyStringsForNullValues;
+            }()
+        )
+    ),
+    _impl(_useEmptyStringsForNullValues ? ToStringWithEmptyStringsForNullValues : ToString) {
+}
+
 template <typename T>
 StringTransformer<T>::StringTransformer(Archive &ar) :
-    BaseType(ar) {
+    StringTransformer(
+        [&ar](void) {
+            // Version
+            std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+            std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+            if(majorVersion != 1 || minorVersion != 0)
+                throw std::runtime_error("Unsupported archive version");
+
+            return StringTransformer(Traits<bool>::deserialize(ar));
+        }()
+    ) {
+}
+
+template <typename T>
+bool StringTransformer<T>::operator==(StringTransformer const &other) const {
+    return _useEmptyStringsForNullValues == other._useEmptyStringsForNullValues;
+}
+
+template <typename T>
+bool StringTransformer<T>::operator!=(StringTransformer const &other) const {
+    return (*this == other) == false;
+}
+
+template <typename T>
+void StringTransformer<T>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<bool>::serialize(ar, _useEmptyStringsForNullValues);
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <typename T>
+// static
+std::string StringTransformer<T>::ToString(T const &value) {
+    return Traits<T>::ToString(value);
+}
+
+template <typename T>
+// static
+std::string StringTransformer<T>::ToStringWithEmptyStringsForNullValues(T const &value) {
+    return ToStringWithEmptyStringsForNullValuesImpl(value, std::integral_constant<bool, Traits<T>::IsNullableType>());
+}
+
+template <typename T>
+// static
+std::string StringTransformer<T>::ToStringWithEmptyStringsForNullValuesImpl(T const &value, std::true_type) {
+    if(Traits<T>::IsNull(value))
+        return "";
+
+    return ToString(value);
+}
+
+template <typename T>
+// static
+std::string StringTransformer<T>::ToStringWithEmptyStringsForNullValuesImpl(T const &, std::false_type) {
+    throw std::runtime_error("This should never be called");
 }
```

Comment:
If this won't ever be called them why do we need it? Is it to make the compiler happy by having both the `std::false_type` and the `std::true_type`?

Reply:
The compiler complains if both methods aren't defined, even though one of them will never be called. I tried to create the declaration (but not the definition), but the compiler seemed to want to see the definition as well.